**IMPORTING NECESSARY LIBRARIES**

In [285]:
from google.colab import drive

In [286]:
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [287]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import tensorflow as tf
from sklearn import preprocessing
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Dropout, BatchNormalization
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score, f1_score, precision_recall_curve, auc
import matplotlib.pyplot as plt
from tensorflow.keras import optimizers



In [288]:
project_path = '/content/drive/My Drive/Colab Notebooks/'

In [289]:
dataset_file = project_path + 'bank.csv'

In [290]:
data = pd.read_csv(dataset_file)

**DATA PREPROCESSING**

In [291]:
data.head()

,RowNumber,CustomerId,Surname,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary,Exited
0,1,15634602,Hargrave,619,France,Female,42,2,0.00,1,1,1,101348.88,1
1,2,15647311,Hill,608,Spain,Female,41,1,83807.86,1,0,1,112542.58,0
2,3,15619304,Onio,502,France,Female,42,8,159660.80,3,1,0,113931.57,1
3,4,15701354,Boni,699,France,Female,39,1,0.00,2,0,0,93826.63,0
4,5,15737888,Mitchell,850,Spain,Female,43,2,125510.82,1,1,1,79084.10,0


In [292]:
data = data.drop(["RowNumber","CustomerId",'Surname'], axis = 1)

In [293]:
X_data = data.iloc[:, :-1]

In [294]:
print(X_data.shape)
print('')
X_data.head()

(10000, 10)



,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary
0,619,France,Female,42,2,0.00,1,1,1,101348.88
1,608,Spain,Female,41,1,83807.86,1,0,1,112542.58
2,502,France,Female,42,8,159660.80,3,1,0,113931.57
3,699,France,Female,39,1,0.00,2,0,0,93826.63
4,850,Spain,Female,43,2,125510.82,1,1,1,79084.10


In [295]:
y_data = data.iloc[:, -1]

In [296]:
print(y_data.shape)
print('')
y_data.head()

(10000,)



0    1
1    0
2    1
3    0
4    0
Name: Exited, dtype: int64

In [297]:
X_train, X_test, y_train, y_test = train_test_split(X_data, y_data, test_size = 0.3, random_state = 7)
X_train, x_val, y_train, y_val = train_test_split(X_train, y_train, test_size=0.25, random_state=1) # 0.25 x 0.8 = 0.2

In [298]:
X_train.head()

,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary
2513,666,France,Male,39,10,0.00,2,1,0,102999.33
1252,693,France,Male,68,4,97705.99,1,1,1,61569.07
7928,711,France,Male,38,4,123345.85,1,1,0,141827.83
5440,690,Germany,Male,43,2,166522.78,1,0,0,119644.59
6662,665,France,Male,36,1,0.00,2,0,1,121505.61


In [299]:
x_val.head()

,CreditScore,Geography,Gender,Age,Tenure,Balance,NumOfProducts,HasCrCard,IsActiveMember,EstimatedSalary
3424,688,Spain,Male,42,5,0.00,2,0,0,197602.29
1081,461,Spain,Female,35,8,0.00,1,1,0,132295.95
6395,741,Germany,Female,36,0,127675.39,2,1,0,74260.16
5670,497,France,Male,32,8,0.00,2,1,0,67364.42
3903,643,Spain,Female,45,9,150840.03,2,1,0,155516.35


In [300]:
X_train = pd.get_dummies (X_train , columns = ['Geography','Gender'], dtype = 'float32', prefix = 'Country')
X_test = pd.get_dummies (X_test , columns = ['Geography','Gender'], dtype = 'float32', prefix = 'Country')
x_val = pd.get_dummies (x_val , columns = ['Geography','Gender'], dtype = 'float32', prefix = 'Country')

In [301]:
X_train_norm = preprocessing.normalize(X_train)
x_val_norm = preprocessing.normalize(x_val)
x_val_norm = x_val_norm.astype('float32')
X_train_norm = X_train_norm.astype('float32')

In [302]:
X_train_norm

array([[6.4659258e-03, 3.7863530e-04, 9.7085976e-05, ..., 0.0000000e+00,
        0.0000000e+00, 9.7085976e-06],
       [6.0005747e-03, 5.8880093e-04, 3.4635352e-05, ..., 0.0000000e+00,
        0.0000000e+00, 8.6588379e-06],
       [3.7826728e-03, 2.0216816e-04, 2.1280859e-05, ..., 0.0000000e+00,
        0.0000000e+00, 5.3202148e-06],
       ...,
       [3.8936413e-03, 2.6036505e-04, 1.1834775e-05, ..., 0.0000000e+00,
        0.0000000e+00, 5.9173876e-06],
       [4.7431975e-03, 2.4128715e-04, 1.2699324e-05, ..., 0.0000000e+00,
        0.0000000e+00, 6.3496618e-06],
       [5.7505076e-03, 4.3251680e-04, 3.9319708e-05, ..., 9.8299270e-06,
        9.8299270e-06, 0.0000000e+00]], dtype=float32)

In [303]:
print(X_train_norm.shape)
print(x_val.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)

(5250, 13)
(1750, 13)
(3000, 13)
(5250,)
(3000,)


**CREATING MODEL**

In [304]:
model = Sequential()

Adding Layers

In [305]:
model.add(Dense(13, input_shape = (13,) , activation = 'relu'))
model.add(Dense(100, activation = 'tanh'))
model.add(Dense(1, activation = 'sigmoid'))

In [306]:
sgd = optimizers.Adam(lr = 0.001)

In [307]:
model.compile(optimizer = sgd, loss = 'categorical_crossentropy', metrics=['accuracy'])

**Training the Model**

In [308]:
model.fit(x_val_norm, y_val, batch_size = 200, epochs = 10, verbose = 1)

Epoch 1/10
9/9 [==============================] - 0s 2ms/step - loss: 2.2275e-08 - accuracy: 0.2137
Epoch 2/10
9/9 [==============================] - 0s 2ms/step - loss: 2.2275e-08 - accuracy: 0.2137
Epoch 3/10
9/9 [==============================] - 0s 2ms/step - loss: 2.2275e-08 - accuracy: 0.2137
Epoch 4/10
9/9 [==============================] - 0s 2ms/step - loss: 2.2275e-08 - accuracy: 0.2137
Epoch 5/10
9/9 [==============================] - 0s 2ms/step - loss: 2.2275e-08 - accuracy: 0.2137
Epoch 6/10
9/9 [==============================] - 0s 2ms/step - loss: 2.2275e-08 - accuracy: 0.2137
Epoch 7/10
9/9 [==============================] - 0s 2ms/step - loss: 2.2275e-08 - accuracy: 0.2137
Epoch 8/10
9/9 [==============================] - 0s 2ms/step - loss: 2.2275e-08 - accuracy: 0.2137
Epoch 9/10
9/9 [==============================] - 0s 2ms/step - loss: 2.2275e-08 - accuracy: 0.2137
Epoch 10/10
9/9 [==============================] - 0s 2ms/step - loss: 2.2275e-08 - accuracy: 0.2137

In [309]:
from sklearn.model_selection import RandomizedSearchCV
from scipy.stats import randint as sp_randint
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score


**Evaluation**

In [310]:
X_test = preprocessing.normalize(X_test)

In [311]:
results = model.evaluate(X_train, y_train.values)

165/165 [==============================] - 0s 883us/step - loss: 2.5091e-08 - accuracy: 0.4611


With Binary crossentropy as the loss function I got a loss of .5, but with categorical the loss is significantly smaller at 2.4e-8. which is also smaller than the mse loss of .158. The number of epochs doesn't change the accuracy or loss nor does the bach size. I then changed the activation function on the hidden layer from tanh to relu and observed no changes to the results. The model was improved by one hot encoding the categorical values of geography and gender as well. Through all iterations the model accuracy was .8. I then trained the model on a validation set which had a higher validation accuracy than training accuracy which suggests overfitting or that there is a problem with the training data. After normalization the accuracy dropped further but the recall and precision values were more accurate with abetter F1 score of .31 instead of 0 suggesting a more generalizable model. Then I changed the activation functions of the input and the hidden layer to the default relu function which increased the accuracy from the new value of .6 to .8 again.

In [312]:
print(model.metrics_names)
print(results)    

['loss', 'accuracy']
[2.5090717059583767e-08, 0.4611428678035736]


In [313]:
y_test

1977    1
3880    0
52      0
2551    0
2246    0
       ..
7477    1
6059    0
133     0
9907    0
3582    0
Name: Exited, Length: 3000, dtype: int64

In [314]:
results = model.evaluate(X_test, y_test.values)

94/94 [==============================] - 0s 1ms/step - loss: 2.4041e-08 - accuracy: 0.2237


In [315]:
print(model.metrics_names)
print(results)    

['loss', 'accuracy']
[2.4040540225200857e-08, 0.22366666793823242]


**Confusion Matrix**

In [316]:
Y_pred_cls= model.predict_classes(X_test, batch_size=200, verbose=1)
print('Accuracy Model1 (Dropout): '+ str(model.evaluate(X_test,y_test)[1]))
print('Recall_score: ' + str(recall_score(y_test,Y_pred_cls)))
print('Precision_score: ' + str(precision_score(y_test, Y_pred_cls)))
print('F-score: ' + str(f1_score(y_test,Y_pred_cls)))
confusion_matrix(y_test, Y_pred_cls)

94/94 [==============================] - 0s 1ms/step - loss: 2.4041e-08 - accuracy: 0.2237
Accuracy Model1 (Dropout): 0.22366666793823242
Recall_score: 0.9487603305785124
Precision_score: 0.1998607242339833
F-score: 0.33016968651136036


array([[  97, 2298],
       [  31,  574]])

The model predicts alot of churn for this bank with a good level of accuracy. They should consider methods of increasing retainment. However, as the F1 score of the model increased from .31 to .33 the ccuracy score decreased significantly from .8 to .22. As a business I'd prefer a more accurate model to a more generalizable one. However, this is the most generalizable model based off its F1 and ROC scores. 